staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go

```diff
@@ -352,17 +356,17 @@ func NewEgressSelector(config *apiserver.EgressSelectorConfiguration) (*EgressSe
 		if err != nil {
 			return nil, fmt.Errorf("failed to create dialer for egressSelection %q: %v", name, err)
 		}
-		cs.egressToDialer[name] = dialerCreator.createDialer()
+		cs.egressToDialer[name] = dialerCreator.createDialerGenerator()
 	}
 	return cs, nil
 }
 
 // Lookup gets the dialer function for the network context.
 // This is configured for the Kubernetes API Server at startup.
-func (cs *EgressSelector) Lookup(networkContext NetworkContext) (utilnet.DialFunc, error) {
+func (cs *EgressSelector) Lookup(networkContext NetworkContext, caller string) (utilnet.DialFunc, error) {
 	if cs.egressToDialer == nil {
 		// The round trip wrapper will over-ride the dialContext method appropriately
 		return nil, nil
 	}
-	return cs.egressToDialer[networkContext.EgressSelectionName], nil
+	return cs.egressToDialer[networkContext.EgressSelectionName](caller), nil
```

Comment:
This really emphasizes that this does not play well with connection/tunnel reuse. What is our plan for reusing a connection for different callers?

Reply:
We already create a separate tunnel (apiserver <-> proxy server) for each request so this should have no impact on tunnel reuse. (Note that proxy server <-> proxy agent tunnel is always reused/multiplexed) @caesarxuchao performed a performance test with reusing the tunnel between apiserver and proxy server and we didn't observe any noticeable performance improvement while a bunch of complexity was added.

Reply:
Which is a great reason to not make that change now. However there is a difference between saying it doesn't make sense now and it doesn't make sense.